yarn-project/noir-protocol-circuits/src/crates/private-kernel-lib/src/private_kernel_init.nr

```diff
@@ -41,43 +38,44 @@ impl PrivateKernelInputsInit {
+        let origin_address_matches = tx_request.origin.eq(call_stack_item.contract_address);
         assert(origin_address_matches, "origin address does not match call stack items contract address");
         //
         // Ensure we are calling the correct initial function in the contract
-        let entry_point_function_matches = tx_request.function_data.hash() == call_stack_item.function_data().hash();
+        let entry_point_function_matches = tx_request.function_data.hash() == call_stack_item.function_data.hash();
         // TODO - Allow special characters in error message.
         // assert(entry_point_function_matches, "tx_request.function_data must match call_stack_item.function_data");
         assert(entry_point_function_matches, "tx_request function_data must match call_stack_item function_data");
         //
         // Ensure we are passing the correct arguments to the function.
-        let args_match = tx_request.args_hash == call_stack_item.public_inputs().args_hash;
+        let args_match = tx_request.args_hash == call_stack_item.public_inputs.args_hash;
         assert(args_match, "noir function args passed to tx_request must match args in the call_stack_item");
     }
 
     fn validate_inputs(self) {
         let call_stack_item = self.private_call.call_stack_item;
-        let function_data = call_stack_item.function_data();
+        let function_data = call_stack_item.function_data;
 
-        let public_inputs = call_stack_item.public_inputs();
+        let public_inputs = call_stack_item.public_inputs;
 
         /* TODO(Mike/David): If we are going to have 3 initial calls on the private call stack,
         * then do we still need the `private_call_stack`
         * despite no longer needing a full `previous_kernel`
         */
 
         assert(function_data.is_private, "Private kernel circuit can only execute a private function");
+        assert(function_data.is_internal == false, "Cannot call an internal function directly");
         assert(public_inputs.call_context.is_delegate_call == false, "Users cannot make a delegatecall");
         assert(public_inputs.call_context.is_static_call == false, "Users cannot make a static call");
         
         // The below also prevents delegatecall/staticcall in the base case
-        assert(public_inputs.call_context.storage_contract_address.eq(call_stack_item.contract_address()), "Storage contract address must be that of the called contract");
+        assert(public_inputs.call_context.storage_contract_address.eq(call_stack_item.contract_address), "Storage contract address must be that of the called contract");
     }
 
-    fn update_end_values(self, public_inputs : &mut KernelCircuitPublicInputsBuilder) {
-        // We only initialized constants member of public_inputs so far. Therefore, there must not be any
-        // new nullifiers or logs as part of public_inputs.
-        
+    // Not needed.
```

Comment:
Let's remove it! As you say since the builder is initialized in-circuit, this will get simplified out by the compiler